runtime/src/main/scala/akka/grpc/scaladsl/headers/PercentEncoding.scala

```diff
@@ -0,0 +1,108 @@
+ * UTF-8 followed by percent-encoding. Percent-encoding is specified in RFC 3986 §2.1, although the form used here has
+ * different restricted characters. When decoding invalid values, implementations MUST NOT error or throw away the
+ * message. At worst, the implementation can abort decoding the status message altogether such that the user would
+ * received the raw percent-encoded form. Alternatively, the implementation can decode valid portions while leaving
+ * broken %-encodings as-is or replacing them with a replacement character (e.g., '?' or the Unicode replacement
+ * character).
+ */
+private[grpc] object PercentEncoding {
+  // Copied with slight adaptations from https://github.com/grpc/grpc-java/blob/79e75bace40cea7e4be72e7dcd1f41c3ad6ee857/api/src/main/java/io/grpc/Status.java#L582
+  object Encoder {
+    private val HexArr: Array[Byte] =
+      Array('0', '1', '2', '3', '4', '5', '6', '7', '8', '9', 'A', 'B', 'C', 'D', 'E', 'F')
+
+    def encode(value: String): String = {
+      val valueBytes = value.getBytes(StandardCharsets.UTF_8)
+      val firstIndexToEscape = valueBytes.indexWhere(isEscapingChar(_))
+      if (firstIndexToEscape == -1)
+        new String(valueBytes)
+      else
+        encodeSlow(valueBytes, firstIndexToEscape)
+    }
+
+    private def isEscapingChar(b: Byte): Boolean = b < ' ' || b >= '~' || b == '%'
+
+    private def encodeSlow(valueBytes: Array[Byte], riArg: Int): String = {
+      var ri = riArg
+      val escapedBytes = new Array[Byte](riArg + ((valueBytes.length - riArg) * 3))
+      // copy over the good bytes
+      if (riArg != 0) System.arraycopy(valueBytes, 0, escapedBytes, 0, riArg)
+      var wi = ri
+
+      while (ri < valueBytes.length) {
+        val b = valueBytes(ri)
+        // Manually implement URL encoding, per the gRPC spec.
+        if (isEscapingChar(b)) {
+          escapedBytes.update(wi, '%')
+          escapedBytes.update(wi + 1, HexArr((b >> 4) & 0xf))
+          escapedBytes.update(wi + 2, HexArr(b & 0xf))
+          wi += 3
+        } else {
+          escapedBytes.update(wi, b)
+          wi += 1
+        }
+        ri += 1
+      }
+      new String(escapedBytes, 0, wi, StandardCharsets.US_ASCII)
+    }
+  }
+
+  // Copied with slight adaptations from https://github.com/grpc/grpc-java/blob/79e75bace40cea7e4be72e7dcd1f41c3ad6ee857/api/src/main/java/io/grpc/Status.java#L626
+  object Decoder {
+    private val TransferEncoding = StandardCharsets.US_ASCII
+
+    def decode(value: String): String =
+      if (value.indexOf('%') > -1)
+        decodeSlow(value)
+      else
+        value
+
+    private def decodeSlow(value: String): String = {
+      val source = value.getBytes(TransferEncoding)
+      val buf = ByteBuffer.allocate(source.length)
+      var i = 0
+      while (i < source.length) {
+        if (source(i) == '%' && i + 2 < source.length) {
+          try {
+            buf.put(Integer.parseInt(new String(source, i + 1, 2, TransferEncoding), 16).toByte)
```

Comment:
Seems extremely costly to do this instead of looking at 2 bytes and create a number?

Reply:
True, reworked according to the suggestion. I ran microbenchmark and the difference for non-contrived input was significant indeed - 40% of speedup